rls/src/test/java/io/grpc/rls/internal/AdaptiveThrottlerTest.java

```diff
@@ -0,0 +1,123 @@
+    assertThat(throttler.getThrottleProbability(fakeTimeProvider.currentTimeNanos()))
+        .isWithin(TOLERANCE).of(0.0f);
+
+    // Request 1, allowed by all.
+    assertThat(throttler.shouldThrottle(0.4f)).isFalse();
+    fakeTimeProvider.advance(1L, TimeUnit.MILLISECONDS);
+    throttler.registerBackendResponse(false);
+
+    assertThat(throttler.requestStat.get(fakeTimeProvider.currentTimeNanos()))
+        .isEqualTo(1L);
+    assertThat(throttler.throttledStat.get(fakeTimeProvider.currentTimeNanos())).isEqualTo(0L);
+    assertThat(throttler.getThrottleProbability(fakeTimeProvider.currentTimeNanos()))
+        .isWithin(TOLERANCE).of(0.0f);
+
+    // Request 2, throttled by backend
+    assertThat(throttler.shouldThrottle(0.4f)).isFalse();
+    fakeTimeProvider.advance(1L, TimeUnit.MILLISECONDS);
+    throttler.registerBackendResponse(true);
+
+    assertThat(throttler.requestStat.get(fakeTimeProvider.currentTimeNanos()))
+        .isEqualTo(2L);
+    assertThat(throttler.throttledStat.get(fakeTimeProvider.currentTimeNanos()))
+        .isEqualTo(1L);
+    assertThat(throttler.getThrottleProbability(fakeTimeProvider.currentTimeNanos()))
+        .isWithin(TOLERANCE)
+        .of(1.0f / 3.0f);
+
+    // Skip half a second (half the duration).
+    fakeTimeProvider.setNow(500L, TimeUnit.MILLISECONDS);
+
+    // Request 3, throttled by backend
+    assertThat(throttler.shouldThrottle(0.4f)).isFalse();
+    fakeTimeProvider.advance(1L, TimeUnit.MILLISECONDS);
+    throttler.registerBackendResponse(true);
+
+    assertThat(throttler.requestStat.get(fakeTimeProvider.currentTimeNanos())).isEqualTo(3L);
+    assertThat(throttler.throttledStat.get(fakeTimeProvider.currentTimeNanos())).isEqualTo(2L);
+    assertThat(throttler.getThrottleProbability(fakeTimeProvider.currentTimeNanos()))
+        .isWithin(TOLERANCE)
+        .of(2.0f / 4.0f);
+
+    // Request 4, throttled by client.
+    assertThat(throttler.shouldThrottle(0.4f)).isTrue();
+    fakeTimeProvider.advance(1L, TimeUnit.MILLISECONDS);
+
+    assertThat(throttler.requestStat.get(fakeTimeProvider.currentTimeNanos())).isEqualTo(4L);
+    assertThat(throttler.throttledStat.get(fakeTimeProvider.currentTimeNanos())).isEqualTo(3L);
+    assertThat(throttler.getThrottleProbability(fakeTimeProvider.currentTimeNanos()))
+        .isWithin(TOLERANCE)
+        .of(3.0f / 5.0f);
+
+    // Skip to the point where only requests 3 and 4 are visible.
+    fakeTimeProvider.setNow(1250L, TimeUnit.MILLISECONDS);
+
+    assertThat(throttler.requestStat.get(fakeTimeProvider.currentTimeNanos())).isEqualTo(2L);
+    assertThat(throttler.throttledStat.get(fakeTimeProvider.currentTimeNanos())).isEqualTo(2L);
+    assertThat(throttler.getThrottleProbability(fakeTimeProvider.currentTimeNanos()))
+        .isWithin(TOLERANCE)
+        .of(2.0f / 3.0f);
+  }
+
+  private static final class FakeTimeProvider implements TimeProvider {
```

Comment:
You can alternatively use `FakeClock.getTimeProvider()` (need `testCompile project(':grpc-core').sourceSets.test.output` though)

Reply:
done.